compiler/shared_ast/typing.ml

```diff
@@ -698,9 +694,9 @@ let infer_types_program prg =
       let rec process_scope_body_expr env = function
         | A.Result e ->
           let e' = wrap ctx (typecheck_expr_bottom_up ctx env) e in
+          wrap ctx (unify ctx e (ty (Bindlib.unbox e'))) ty_out;
```

Comment:
I am not really sure this is correct though: if some unionfinds ever get generated at `unbox` time, they will be duplicated and we unify on an instance that will get dropped, leaving another instance in the marked AST.

This would also mean that part of the typing work gets done twice

Reply:
Remember me why we don't simply let a `StructuredError` generated at unification failure go up to get caught in `driver.ml` ?

Reply:
Hm initially there were difficulties with reporting the proper expressions where the error occured (and not a sub/surrounding one) with `top_down` and `bottom_up` wrapping each other's calls. But indeed that's no longer the case so that could probably be simplified!

This wouldn't, however, solve the issue of typing error getting delayed more than we'd like :thinking: